saffron/src/query.rs

```diff
@@ -0,0 +1,292 @@
+use std::marker::PhantomData;
+
+use crate::utils::decode_from_field_elements;
+use ark_ff::PrimeField;
+use ark_poly::{EvaluationDomain, Evaluations, Radix2EvaluationDomain};
+use rayon::prelude::*;
+use thiserror::Error;
+
+#[derive(Clone, Debug)]
+/// Represents the bytes a user query
+pub struct QueryBytes {
+    pub start: usize,
+    pub len: usize,
+}
+
+#[derive(Copy, Clone, PartialEq, PartialOrd, Eq, Ord, Debug)]
+/// We store the data in a vector of vector of field element
+/// The inner vector represent polynomials
+pub struct FieldElt {
+    /// the index of the polynomial the data point is attached too
+    pub poly_index: usize,
+    /// the index of the root of unity the data point is attached too
+    pub eval_index: usize,
+    domain_size: usize,
+    n_polys: usize,
+}
+/// Represents a query in term of Field element
+#[derive(Debug)]
+pub struct QueryField<F> {
+    pub start: FieldElt,
+    /// how many bytes we need to trim from the first chunk
+    /// we get from the first field element we decode
+    pub leftover_start: usize,
```

Comment:
I guess this is going to be removed? It seems over-complicated, as we are supposed to talk in terms of field elements only, no?

Reply:
If your looking to read a slice of bytes from encoded data, the slice might start in the middle of a field element. This is used to trim the excess bytes in that case.

This is something that doesn't affect proving at all, which is only communicated in terms of field elems. It's used by the user to parse the returned values, they can construct these limit values themselves without interacting with anyone else.

Reply:
Given that the demo is supposed to show the use of saffron on bytes, it seems like an important thing to include.  Do you disagree, @dannywillems ?  I will resolve if you're okay with it.

Reply:
I think the demo is supposed to use field elements, in reference to https://github.com/o1-labs/proof-systems/tree/dw/demo-different-encodings-saffron/saffron/src/encoding

Reply:
Sorry, I am a little unclear on what you mean, @svv232.  Indeed, the saffron demo uses field elements, but user's data (for example, a lorem ipsum file) will not be preencoded to field elements.  Performing reads on the underlying byte data in the dense representation (the only representation currently implemented) would require this logic.  Even if we expect to use a sparse representation for some demos (which seems unlikely tbh), this would still be needed for the dense representation.

Reply:
For example, if I wanted to query bytes 4-8, I would need to perform this kind of logic.